hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/PulsarSource.java

```
@@ -0,0 +1,292 @@
+        .option("topics", topicName)
+        .option("startingOffsets", startingOffsetStr)
+        .option("endingOffsets", endingOffsetStr)
+        .load();
+
+    return Pair.of(Option.of(transform(sourceRows)), endingOffsetStr);
+  }
+
+  @Override
+  public void onCommit(String lastCheckpointStr) {
+    MessageId latestConsumedOffset = JsonUtils.topicOffsets(lastCheckpointStr).apply(topicName);
+    ackOffset(latestConsumedOffset);
+  }
+
+  private Dataset<Row> transform(Dataset<Row> rows) {
+    return rows.drop(PULSAR_META_FIELDS);
+  }
+
+  private Pair<MessageId, MessageId> computeOffsets(Option<String> lastCheckpointStrOpt, long sourceLimit) {
+    MessageId startingOffset = decodeStartingOffset(lastCheckpointStrOpt);
+    MessageId endingOffset = fetchLatestOffset();
+
+    if (endingOffset.compareTo(startingOffset) < 0) {
+      String message = String.format("Ending offset (%s) is preceding starting offset (%s) for '%s'",
+          endingOffset, startingOffset, topicName);
+      throw new HoodieException(message);
+    }
+
+    // TODO support capping the amount of records fetched
+    Long maxRecordsLimit = computeTargetRecordLimit(sourceLimit, props);
+
+    return Pair.of(startingOffset, endingOffset);
+  }
+
+  private MessageId decodeStartingOffset(Option<String> lastCheckpointStrOpt) {
+    return lastCheckpointStrOpt
+        .map(lastCheckpoint -> JsonUtils.topicOffsets(lastCheckpoint).apply(topicName))
+        .orElseGet(() -> {
+          Config.OffsetAutoResetStrategy autoResetStrategy = Config.OffsetAutoResetStrategy.valueOf(
+              props.getString(Config.PULSAR_SOURCE_OFFSET_AUTO_RESET_STRATEGY.key(),
+                  Config.PULSAR_SOURCE_OFFSET_AUTO_RESET_STRATEGY.defaultValue().name()));
+
+          switch (autoResetStrategy) {
+            case LATEST:
+              return fetchLatestOffset();
+            case EARLIEST:
+              return MessageId.earliest;
+            case FAIL:
```

whats the purpose of "FAIL". we don't have one for Kafka right.

---

but don't we have a default value as "LATEST" ?

---

We do. But setting it `FAIL` allows you to fail-fast instead of skipping the data w/ `LATEST` (for ex, you can recover by manually setting the checkpoint in that case)